platforms/core-runtime/base-services/src/main/java/org/gradle/util/Path.java

```diff
@@ -74,6 +74,8 @@ private static Path parsePath(String path) {
     private volatile String fullPath;
 
     private Path(String[] segments, boolean absolute) {
+        assert !(segments.length == 0 && !absolute) : "Empty relative paths are forbidden";
```

Comment:
💅   I find the current condition harder to read, but maybe it's just me.

This reads better for me as "either it's absolute, or if not, then must have segments"
```suggestion
        assert absolute || segments.length != 0 : "Empty relative paths are forbidden";
```

Reply:
❔  Why not throw an exception instead?

Also, can we / should we write a test that triggers the condition to validate it is in place?

Reply:
> Why not throw an exception instead?

The constructor is private, so we are not concerned with _users_ going and creating empty relative paths. The public factory method for this class, `Path.path(String)` _does_ forbid null or empty strings -- which is effectively equivalent.

At first when writing optimizations/tests for this class I did not know empty relative paths were forbidden. I added this assertion after discovering that. I find assertions useful as codified assertions -- documenting assumptions we make when writing code, and verifying them with code -- but without any runtime overhead. Assertions _are_ executed when running tests but not during production.

In effect I am not concerned with users going and making empty relative paths, but am concerned with developers working on this class breaking assumptions that this class made already (empty relative paths don't exist)

Reply:
As for the assertion structure, I am not tied to this way. 

The current assertion reads "ensure the input is not empty and relative", the form you suggested I read as "ensure the input is absolute or non-empty".

When using the `||` formation, you have to go and think of the whole set of paths, minus the ones that are non-empty, minus the set that are absolute, and the remaining are forbidden. 

The `&&` formation, we tell you outright the forbidden set of inputs are those which are empty and relative.

In any case I don't really care that much. I've change it to what you suggested, but left my message `Empty relative paths are forbidden`

Reply:
I see now how you initially mirrored the words of the error message in the boolean condition. I missed that and focused on describing the entirely of the invariant set in a more concise way. The latter seems to be less useful, when we have exactly one unsupported invariant.

I indicated the initial comment with 💅  to let you decide which one to choose, as I only wanted you to consider an additional suggestion.

Reply:
Given that, I have gone back to the old assertion